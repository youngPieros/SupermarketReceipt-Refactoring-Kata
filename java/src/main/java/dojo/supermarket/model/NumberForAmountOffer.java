package dojo.supermarket.model;

import java.util.Map;

public class NumberForAmountOffer extends SingleProductOffer {

    private final int numberOfAmountOffer;
    private final double amount;

    public NumberForAmountOffer(Product product, int numberOfAmountOffer, double amount) {
        super(product);
        this.numberOfAmountOffer = numberOfAmountOffer;
        this.amount = amount;
    }

    @Override
    public Discount calculateDiscount(Map<Product, Double> productQuantities) {
        int quantity = productQuantities.get(this.getProduct()).intValue();
        int quantityNumberWithDiscount = (quantity / numberOfAmountOffer * numberOfAmountOffer);
        double discount = quantityNumberWithDiscount * (this.getProduct().getPrice() - this.amount / numberOfAmountOffer);
        return Discount.createDiscount(this.getProduct(), this.numberOfAmountOffer + " for " + this.amount, discount);
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return productQuantities.get(getProduct()) >= numberOfAmountOffer;
    }

}
