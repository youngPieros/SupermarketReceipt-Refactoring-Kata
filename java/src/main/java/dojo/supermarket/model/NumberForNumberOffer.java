package dojo.supermarket.model;

import java.util.Map;

public class NumberForNumberOffer extends SingleProductOffer {

    private final int numberOfBoughtProducts;
    private final int numberOfPaidProducts;

    public NumberForNumberOffer(Product product, int numberOfPaidProducts, int numberOfBoughtProducts) {
        super(product);
        this.numberOfBoughtProducts = numberOfBoughtProducts;
        this.numberOfPaidProducts = numberOfPaidProducts;
    }

    @Override
    protected Discount calculateDiscount(Map<Product, Double> productQuantities) {
        int quantity = productQuantities.get(this.getProduct()).intValue();
        int quantityNumberWithDiscount = quantity / numberOfBoughtProducts;
        double discount = quantityNumberWithDiscount * (this.numberOfBoughtProducts - this.numberOfPaidProducts) * this.getProduct().getPrice();
        String description = numberOfBoughtProducts + " for " + numberOfPaidProducts;
        return Discount.createDiscount(this.getProduct(), description, discount);
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return productQuantities.get(getProduct()) >= numberOfBoughtProducts;
    }
}
