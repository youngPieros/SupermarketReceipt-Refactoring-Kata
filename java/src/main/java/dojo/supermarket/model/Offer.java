package dojo.supermarket.model;
import java.util.*;

public interface Offer {
    Discount calculateDiscount(Map<Product, Double> productQuantities);

    Double getIncludedProductDiscountNumber(Map<Product, Double> productQuantities);
}
