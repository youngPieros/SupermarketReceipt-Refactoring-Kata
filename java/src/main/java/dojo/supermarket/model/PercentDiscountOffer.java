package dojo.supermarket.model;

import java.util.Map;

public class PercentDiscountOffer extends AbstractOffer {

    private final Product product;
    private final double amountPercent;

    public PercentDiscountOffer(Product product, double amountPercent) {
        this.product = product;
        this.amountPercent = amountPercent;
    }


    @Override
    public Discount calculateDiscount(Map<Product, Double> productQuantities) {
        double quantityNumberWithDiscount = getIncludedProductDiscountNumber(productQuantities);
        double discount = quantityNumberWithDiscount * product.getPrice() * amountPercent / 100;
        String description = this.amountPercent + "% off";
        return Discount.createDiscount(description, discount);
    }

    @Override
    public Double getIncludedProductDiscountNumber(Map<Product, Double> productQuantities) {
        return productQuantities.get(product);
    }
}
