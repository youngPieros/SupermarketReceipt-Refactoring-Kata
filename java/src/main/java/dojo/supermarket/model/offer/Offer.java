package dojo.supermarket.model.offer;
import dojo.supermarket.model.discount.Discount;
import dojo.supermarket.model.product.Product;

import java.util.*;

public interface Offer {
    public List<Discount> calculateDiscounts(Map<Product, Double> productQuantities);

    public boolean isApplicable(Map<Product, Double> productQuantities);
}
