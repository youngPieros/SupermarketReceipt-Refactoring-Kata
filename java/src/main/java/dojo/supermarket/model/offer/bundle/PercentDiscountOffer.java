package dojo.supermarket.model.offer.bundle;

import dojo.supermarket.model.discount.Discount;
import dojo.supermarket.model.product.Product;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class PercentDiscountOffer extends BundleProductOffer {

    private final double amountPercent;

    public PercentDiscountOffer(ArrayList<Product> products, double amountPercent) {
        super(products);
        this.amountPercent = amountPercent;
    }

    @Override
    public ArrayList<Discount> calculateBundleDiscounts(Map<Product, Double> productQuantities) {
        ArrayList<Product> products = getProducts();
        ArrayList<Discount> discounts = new ArrayList<>();
        int quantityNumberOfDiscountProduct = getDiscountQuantities(productQuantities).get(products.get(0)).intValue();
        for (Product product: products) {
            String description = this.amountPercent + "% off";
            double discount = quantityNumberOfDiscountProduct * product.getPrice()  * amountPercent / 100;
            discounts.add(Discount.createDiscount(product, description, discount));
        }
        return discounts;
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return true;
    }

    @Override
    protected Map<Product, Double> getDiscountQuantities(Map<Product, Double> productQuantities) {
        Map<Product, Double> discountQuantities = new HashMap<>();
        ArrayList<Product> products = getProducts();
        int quantityNumberOfDiscountProduct = productQuantities.get(products.get(0)).intValue();
        for (Product product: products) {
            int quantityProduct = productQuantities.get(product).intValue();
            if (quantityNumberOfDiscountProduct > quantityProduct)
                quantityNumberOfDiscountProduct = quantityProduct;
        }
        for (Product product: products)
            discountQuantities.put(product, (double) quantityNumberOfDiscountProduct);
        return discountQuantities;
    }
}
