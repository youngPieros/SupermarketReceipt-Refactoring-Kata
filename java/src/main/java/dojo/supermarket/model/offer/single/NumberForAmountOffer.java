package dojo.supermarket.model.offer.single;

import dojo.supermarket.model.discount.Discount;
import dojo.supermarket.model.product.Product;

import java.util.Map;

public class NumberForAmountOffer extends SingleProductOffer {

    private final int numberOfAmountOffer;
    private final double amount;

    public NumberForAmountOffer(Product product, int numberOfAmountOffer, double amount) {
        super(product);
        this.numberOfAmountOffer = numberOfAmountOffer;
        this.amount = amount;
    }

    @Override
    public Discount calculateDiscount(Map<Product, Double> productQuantities) {
        int quantity = productQuantities.get(this.getProduct()).intValue();
        int quantityNumberWithDiscount = (quantity / numberOfAmountOffer * numberOfAmountOffer);
        double discount = quantityNumberWithDiscount * (this.getProduct().getPrice() - this.amount / numberOfAmountOffer);
        return Discount.createDiscount(this.getProduct(), this.numberOfAmountOffer + " for " + this.amount, discount);
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return productQuantities.get(getProduct()) >= numberOfAmountOffer;
    }

}
