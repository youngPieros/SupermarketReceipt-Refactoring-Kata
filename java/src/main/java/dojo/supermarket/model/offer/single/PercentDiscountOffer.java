package dojo.supermarket.model.offer.single;

import dojo.supermarket.model.discount.Discount;
import dojo.supermarket.model.product.Product;

import java.util.Map;

public class PercentDiscountOffer extends SingleProductOffer {

    private final double amountPercent;

    public PercentDiscountOffer(Product product, double amountPercent) {
        super(product);
        this.amountPercent = amountPercent;
    }

    @Override
    protected Discount calculateDiscount(Map<Product, Double> productQuantities) {
        double quantityNumberWithDiscount = productQuantities.get(getProduct());
        double discount = quantityNumberWithDiscount * this.getProduct().getPrice() * amountPercent / 100;
        String description = this.amountPercent + "% off";
        return Discount.createDiscount(this.getProduct(), description, discount);
    }

    @Override
    protected boolean isApplicableOffer(Map<Product, Double> productQuantities) {
        return true;
    }
}
