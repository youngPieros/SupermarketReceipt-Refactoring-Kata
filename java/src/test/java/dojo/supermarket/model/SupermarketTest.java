package dojo.supermarket.model;

import dojo.supermarket.ReceiptPrinter;
import dojo.supermarket.model.offer.single.NumberForAmountOffer;
import dojo.supermarket.model.offer.single.NumberForNumberOffer;
import dojo.supermarket.model.offer.Offer;
import dojo.supermarket.model.offer.single.PercentDiscountOffer;
import dojo.supermarket.model.product.Product;
import dojo.supermarket.model.product.ProductUnit;
import dojo.supermarket.model.receipt.Receipt;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;

public class SupermarketTest {
    private SupermarketCatalog catalog;
    private Teller teller;
    private ShoppingCart cart;
    private Product toothbrush;
    private Product rice;
    private Product apples;
    private Product cherryTomatoes;

    @BeforeEach
    public void setUp() {
        catalog = new FakeCatalog();
        teller = new Teller(catalog);
        cart = new ShoppingCart();

        toothbrush = new Product("toothbrush", 0.99, ProductUnit.Each);
        catalog.addProduct(toothbrush, 0.99);
        rice = new Product("rice", 2.99, ProductUnit.Each);
        catalog.addProduct(rice, 2.99);
        apples = new Product("apples", 1.99, ProductUnit.Kilo);
        catalog.addProduct(apples, 1.99);
        cherryTomatoes = new Product("cherry tomato box", 0.69, ProductUnit.Each);
        catalog.addProduct(cherryTomatoes, 0.69);

    }

    @Test
    public void an_empty_shopping_cart_should_cost_nothing() {
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void one_normal_item() {
        cart.addItem(toothbrush);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void two_normal_items() {
        cart.addItem(toothbrush);
        cart.addItem(rice);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void buy_two_get_one_free() {
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        Offer offer = new NumberForNumberOffer(toothbrush, 2, 3);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void buy_two_get_one_free_but_insufficient_in_basket() {
        cart.addItem(toothbrush);
        Offer offer = new NumberForNumberOffer(toothbrush, 2, 3);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }
    @Test
    public void buy_five_get_one_free() {
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        Offer offer = new NumberForNumberOffer(toothbrush, 2, 3);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void loose_weight_product() {
        cart.addItemQuantity(apples, .5);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void percent_discount() {
        cart.addItem(rice);
        Offer offer = new PercentDiscountOffer(rice, 10.0);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void xForY_discount() {
        cart.addItem(cherryTomatoes);
        cart.addItem(cherryTomatoes);
        Offer offer = new NumberForAmountOffer(cherryTomatoes, 2, 0.99);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void xForY_discount_with_insufficient_in_basket() {
        cart.addItem(cherryTomatoes);
        Offer offer = new NumberForAmountOffer(cherryTomatoes, 2, 0.99);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount() {
        cart.addItemQuantity(apples, 5);
        Offer offer = new NumberForAmountOffer(apples, 5, 6.99);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount_withSix() {
        cart.addItemQuantity(apples, 6);
        Offer offer = new NumberForAmountOffer(apples, 5, 5.99);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount_withSixteen() {
        cart.addItemQuantity(apples, 16);
        Offer offer = new NumberForAmountOffer(apples, 5, 7.99);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void FiveForY_discount_withFour() {
        cart.addItemQuantity(apples, 4);
        Offer offer = new NumberForAmountOffer(apples, 5, 8.99);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

    @Test
    public void Bundle_discount_test() {
        cart.addItemQuantity(apples, 10);
        cart.addItemQuantity(toothbrush, 5);
        ArrayList<Product> bundleOffers = new ArrayList<>();
        bundleOffers.add(apples);
        bundleOffers.add(toothbrush);
        Offer offer = new dojo.supermarket.model.offer.bundle.PercentDiscountOffer(bundleOffers, 10);
        teller.addSpecialOffer(offer);
        Receipt receipt = teller.toReceipt(cart);
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }
}
