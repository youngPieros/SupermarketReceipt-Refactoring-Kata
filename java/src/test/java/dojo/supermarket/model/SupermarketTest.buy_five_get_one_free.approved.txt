toothbrush                          4.95
  0.99 * 5
3 for 2(toothbrush)                -0.99

Total:                              3.96
